File: Python/Card Game.py
import random

# --- Welcome and rules
print()
print("Welcome to The Card Game!")
print("In this game, each player draws a card, the cards are compared and the winner takes both cards.")
print()
input("Press enter to continue.")
print()

# --- Authentication system

# Gets txt file of player details as a list as ["name, password", "name, password"...]
with open("player_list.csv") as f:
    player_list = f.read().splitlines()


# - Define an authenticator


def auth(name, password):
    """Authorise player names and passwords."""
    # Formats player details correctly
    details = name + ", " + password
    match = 0

    # Searches player_list for details. If found, make match = 1
    for x in player_list:
        if x == details:
            match = 1

    # If no match was found
    if match != 1:
        print("Sorry, " + name + ", your username or password was incorrect.")
        input("Press enter to exit.")
        quit()

    print()
    print("Welcome " + name + "!")
    print()


# - Authenticate players
p1Name = input("Player 1 please enter your name: ")
p1Pass = input("And your password: ")

auth(p1Name, p1Pass)

p2Name = input("Player 2 please enter your name: ")
p2Pass = input("And your password: ")

# Check that player 2 is a different person
if p2Name == p1Name:
    print("Sorry, " + p2Name + ", but that's the same account as player 1.")
    input("Press enter to exit.")
    quit()

auth(p2Name, p2Pass)


# --- The game

# Get deck as a list as ["colour number", ...]
with open("deck.txt") as f:
    deck = f.read().splitlines()

random.shuffle(deck)

colourDict = {
    'Red Black': 'Red',
    'Black Red': 'Red',
    'Yellow Red': 'Yellow',
    'Red Yellow': 'Yellow',
    'Black Yellow': 'Black',
    'Yellow Black': 'Black'
}


def win(card_list, player):
    """Append cards to winner's stack and report win."""
    # Appends both cards to the winner's card stack
    card_list.append(p1ActiveCard)
    card_list.append(p2ActiveCard)
    print(player + " won that hand!")
    print()
    input("Press enter to continue.")
    print()


def colour_compare(colour1, colour2):
    """Compare colours and declare winner."""
    colour = colour1 + " " + colour2
    # Compares concatenated colour string with dictionary to get winning colour
    colour_win = colourDict.get(colour)
    if colour_win == colour1:
        win(p1_cards, p1Name)
    else:
        win(p2_cards, p2Name)


# Initialises player's card stacks as empty
p1_cards = []
p2_cards = []

input("Let's begin the game!")
print()

# Loop until deck is empty
while len(deck) > 0:
    # Both players take the top card
    p1ActiveCard = deck[0]
    del deck[0]
    p2ActiveCard = deck[0]
    del deck[0]

    print(p1Name + " drew a " + p1ActiveCard)
    print(p2Name + " drew a " + p2ActiveCard)
    input("Press enter to continue.")
    print()

    p1Colour = p1ActiveCard.split(" ")[0]
    p2Colour = p2ActiveCard.split(" ")[0]

    # If colours are the same, largest number wins
    if p1Colour == p2Colour:
        p1Number = int(p1ActiveCard.split(" ")[1])
        p2Number = int(p2ActiveCard.split(" ")[1])

        if p1Number > p2Number:
            win(p1_cards, p1Name)
        else:
            win(p2_cards, p2Name)

    # If colours are different, call function
    else:
        colour_compare(p1Colour, p2Colour)

print("All cards have been drawn!")
input("The winner is...")
print()

if len(p1_cards) > len(p2_cards):
    winner = p1Name
    winNum = len(p1_cards)
    win_cards = p1_cards
else:
    winner = p2Name
    winNum = len(p2_cards)
    win_cards = p2_cards

print(winner + "! With " + str(winNum) + " cards!")
print()

input("They had these cards:")
print()

for i in win_cards:
    print(i)

print()

# Append score & player to scores.txt
with open("scores.txt", "a") as f:
    f.write(str(winNum) + " " + winner + "\n")

# Get scores.txt as list called scores_all[] in the form ["score name", ...]
with open("scores.txt") as f:
    scores_all = f.read().splitlines()

# Python will just sort the scores for me. Thanks guys!
scores_high = sorted(scores_all, reverse=True)

input("These were the high scores:")
print()
for i in range(5):
    print(scores_high[i])

print()
print("Thank you, " + p1Name + " and " + p2Name + ", for playing The Card Game!")
input("Goodbye!")

File: Java/Card-Game/src/CardGame.java
import java.io.*;
import java.util.*;

class Card {
    private final String colour;
    private final int number;

    Card(String colour, int number) {
        this.colour = colour;
        this.number = number;
    }

    public String getColour() { return colour; }
    public int getNumber() { return number; }
    public String displayName() { return this.colour + " " + this.number; }
}

class Score {
    private final String name;
    private final int number;

    Score(String name, int number) {
        this.name = name;
        this.number = number;
    }

    public int getNumber() { return number; }
    public String displayName() { return this.name + " - " + this.number; }
}

public class CardGame {

    static String p1Name;
    static String p2Name;

    static Card p1ActiveCard;
    static Card p2ActiveCard;

    // Create ArrayLists for players' card stacks
    static ArrayList<Card> p1Cards = new ArrayList<>();
    static ArrayList<Card> p2Cards = new ArrayList<>();

    // Create scanner object to read input
    static Scanner inputScanner = new Scanner(System.in);


    public static void authenticate(String name, String password) {
        String playerDetails = name + "," + password;
        boolean match = false;

        // Search through player_list.csv to find playerDetails
        try {
            File playerList = new File("player_list.csv");
            Scanner playerListReader = new Scanner(playerList);

            while (playerListReader.hasNextLine()) {
                String fileDetails = playerListReader.nextLine();
                if (fileDetails.equals(playerDetails)) {
                    match = true;
                }
            }
        }
        catch (FileNotFoundException e) {
            System.out.println("player_list.csv not found.");
            e.printStackTrace();
        }

        // If no match was found, quit program
        if (!match) {
            System.out.println();
            System.out.println("Sorry, " + name + ", your username or password was incorrect.");
            System.out.println("Press enter to exit");
            inputScanner.nextLine();
            System.exit(0);
        }

        System.out.println();
        System.out.println("Welcome, " + name + "!");
        System.out.println();

    }

    public static void addPlayer() {
        System.out.println();
        System.out.print("Please enter the name of the new player: ");
        String name = inputScanner.nextLine();
        System.out.print("Please enter the password: ");
        String password = inputScanner.nextLine();

        String newPlayer = name + "," + password;

        // Add newPlayer to player_list.csv
        try {
            // True enables append mode
            PrintWriter printWriter = new PrintWriter(new FileWriter("player_list.csv", true));
            printWriter.println(newPlayer);
            printWriter.close();
        }
        catch (IOException e) {
            System.out.println("Failed to write to player_list.csv.");
            e.printStackTrace();
        }

        System.out.println();
        System.out.println("Press enter to log in.");
        inputScanner.nextLine();
    }

    public static void winHand(String name, ArrayList<Card> cardStack) {
        // Add active Cards to player's card stack
        cardStack.add(p1ActiveCard);
        cardStack.add(p2ActiveCard);

        System.out.println(name + " won that hand!");
        System.out.println();
        System.out.println("Press enter to continue.");
        inputScanner.nextLine();
    }

    public static void compare(Card card1, Card card2) {
        // If colours are equal, higher number wins
        if (card1.getColour().equals(card2.getColour())) {
            if (card1.getNumber() > card2.getNumber()) {
                winHand(p1Name, p1Cards);
            } else {
                winHand(p2Name, p2Cards);
            }
        } else {
            // If numbers are the same, compare colours to find winner
            String colour = card1.getColour() + " " + card2.getColour();
            String winColour = "";

            if (colour.equals("Red Black") | colour.equals("Black Red")) {
                winColour = "Red";
            } else if (colour.equals("Yellow Red") | colour.equals("Red Yellow")) {
                winColour = "Yellow";
            } else if (colour.equals("Black Yellow") | colour.equals("Yellow Black")) {
                winColour = "Black";
            }

            // Use winColour to find winning player
            if (winColour.equals(card1.getColour())) {
                winHand(p1Name, p1Cards);
            } else if (winColour.equals(card2.getColour())) {
                winHand(p2Name, p2Cards);
            } else {
                // If neither player wins, print error message
                System.out.println("Unhandled winColour exception in CardGame.compare().");
            }
        }
    }


    public static void main(String[] args) {
        System.out.println("Welcome to The Card Game!");
        System.out.println("In this game, each player draws a card, the cards are compared and the winner takes both cards.");
        System.out.println();
        System.out.println("Press 1 to add a new player. Press enter to log in.");

        // Call addPlayer() if input == "1"
        if (inputScanner.hasNextLine()) {
            if (inputScanner.nextLine().equals("1")) {
                addPlayer();
            }
        }

        System.out.print("Player 1, please enter your username: ");
        p1Name = inputScanner.nextLine();

        System.out.print("Player 1, please enter your password: ");
        String p1Pass = inputScanner.nextLine();

        authenticate(p1Name, p1Pass);

        System.out.print("Player 2, please enter your username: ");
        p2Name = inputScanner.nextLine();

        System.out.print("Player 2, please enter your password: ");
        String p2Pass = inputScanner.nextLine();

        // If same account, quit program
        if (p1Name.equals(p2Name)) {
            System.out.println("Sorry, " + p2Name + ", but that's the same account as player 1.");
            System.out.println("Press enter to exit");
            inputScanner.nextLine();
            System.exit(0);
        }

        authenticate(p2Name, p2Pass);

        // Create ArrayList of Cards to allow for shuffling later
        ArrayList<Card> deckList = new ArrayList<>();

        // Read deck.csv into deckList
        try {
            File deckFile = new File("deck.csv");
            Scanner deckScanner = new Scanner(deckFile);

            while (deckScanner.hasNextLine()) {
                String line = deckScanner.nextLine();
                deckList.add(new Card(line.split(",")[0], Integer.parseInt(line.split(",")[1])));
            }
        }
        catch (FileNotFoundException e) {
            System.out.println("deck.csv not found.");
            e.printStackTrace();
        }

        // Shuffle deckList and create normal array called deck
        Collections.shuffle(deckList);
        Card[] deck = deckList.toArray(new Card[0]);

        System.out.println("Press enter to start.");
        inputScanner.nextLine();
        System.out.println("Let's begin the game!");
        inputScanner.nextLine();

        int handNum = 1;

        // Loop over deck and compare top 2 Cards every time
        for (int i = 0; i < deck.length; i = i + 2) {
            System.out.println("Hand: " + handNum);
            p1ActiveCard = deck[i];
            p2ActiveCard = deck[i + 1];

            System.out.println(p1Name + " drew a " + p1ActiveCard.getColour() + " " + p1ActiveCard.getNumber());
            System.out.println(p2Name + " drew a " + p2ActiveCard.getColour() + " " + p2ActiveCard.getNumber());
            System.out.println();
            System.out.println("Press enter to continue.");
            inputScanner.nextLine();

            // Compare Cards to find winner
            compare(p1ActiveCard, p2ActiveCard);

            handNum++;
        }

        System.out.println("All cards have been drawn!");
        System.out.println("The winner is...");
        inputScanner.nextLine();

        // Create variables for winner
        String winner;
        int winNum;
        ArrayList<Card> winCards;

        // Find winner from length of card stacks
        if (p1Cards.size() > p2Cards.size()) {
            winner = p1Name;
            winNum = p1Cards.size();
            winCards = p1Cards;
        } else {
            winner = p2Name;
            winNum = p2Cards.size();
            winCards = p2Cards;
        }

        String winScore = winner + "," + winNum;

        System.out.println(winner + "! With " + winNum + " cards!");
        System.out.println();
        System.out.println("They had these cards:");
        inputScanner.nextLine();

        // Print all Cards
        for (Card card : winCards) {
            System.out.println(card.displayName());
        }

        System.out.println();

        // Write score to file
        try {
            // True enables append mode
            PrintWriter printWriter = new PrintWriter(new FileWriter("scores.csv", true));
            printWriter.println(winScore);
            printWriter.close();
        }
        catch (IOException e) {
            System.out.println("Failed to write to player_list.csv.");
            e.printStackTrace();
        }

        // Create scoresList as ArrayList to allow for sorting later
        ArrayList<Score> scoresList = new ArrayList<>();

        // Read scores from file into scoresList
        try {
            File scoreFile = new File("scores.csv");
            Scanner scoreScanner = new Scanner(scoreFile);

            while (scoreScanner.hasNextLine()) {
                String line = scoreScanner.nextLine();
                scoresList.add(new Score(line.split(",")[0], Integer.parseInt(line.split(",")[1])));
            }
        }
        catch (FileNotFoundException e) {
            System.out.println("scores.csv not found.");
            e.printStackTrace();
        }

        // Sort scoresList by number
        scoresList.sort(Comparator.comparing(Score::getNumber));
        Collections.reverse(scoresList);
        Score[] scores = scoresList.toArray(new Score[0]);

        System.out.println("These are the high scores:");
        inputScanner.nextLine();

        // Print top 5 high scores if available
        if (scores.length >= 5) {
            for (int i = 0; i < 5; i++) {
                System.out.println(scores[i].displayName());
            }
        } else {
            System.out.println("There are not enough scores to display.");
        }

        System.out.println();
        System.out.println("Press enter to finish.");
        inputScanner.nextLine();
        System.out.println("Thank you, " + p1Name + " and " + p2Name + ", for playing The Card Game!");
        inputScanner.nextLine();
    }

}
